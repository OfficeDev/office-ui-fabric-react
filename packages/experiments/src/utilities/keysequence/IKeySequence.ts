import { ModifierKeyCodes } from 'office-ui-fabric-react/lib/Utilities';
import { ktpPrefix, ktpSeparator } from '../keytip/KeytipUtils';

export interface IKeySequence {
  keys: string[];
}

// TODO: how to handle more than one modifier key as a transition sequence
// i.e. Alt-Shift-H
export interface IKeytipTransitionKey {
  key: string;
  modifierKey?: ModifierKeyCodes;
}

export interface IKeytipTransitionSequence {
  keys: IKeytipTransitionKey[];
}

/**
 * Tests for equality between two IKeySequences
 * @param seq1 - First IKeySequence
 * @param seq2 - Second IKeySequence
 */
export function keySequencesAreEqual(seq1: IKeySequence, seq2: IKeySequence): boolean {
  let keyCodes1 = seq1.keys.join();
  let keyCodes2 = seq2.keys.join();
  return keyCodes1 === keyCodes2;
}

/**
 *
 * @param seq1
 * @param seq2
 */
export function transitionKeySequencesAreEqual(seq1: IKeytipTransitionSequence, seq2: IKeytipTransitionSequence): boolean {
  let keys1 = seq1.keys;
  let keys2 = seq2.keys;
  if (keys1.length !== keys2.length) {
    return false;
  }
  for (let i = 0; i < keys1.length; i++) {
    if (keys1[i].key !== keys2[i].key || keys1[i].modifierKey !== keys2[i].modifierKey) {
      return false;
    }
  }
  return true;
}

/**
 *
 * @param sequences
 * @param seq
 */
export function keySequencesContain(sequences: IKeySequence[], seq: IKeySequence): boolean {
  for (let i = 0; i < sequences.length; i++) {
    if (keySequencesAreEqual(sequences[i], seq)) {
      return true;
    }
  }
  return false;
}

export function transitionKeySequencesContain(sequences: IKeytipTransitionSequence[], seq: IKeytipTransitionSequence): boolean {
  for (let i = 0; i < sequences.length; i++) {
    if (transitionKeySequencesAreEqual(sequences[i], seq)) {
      return true;
    }
  }
  return false;
}

/**
 * Method returns true if the key squence of the object with minimum length is in the other key sequence.
 * If the minium length is zero, then it will default to false.
 * @param seq1
 * @param seq2
 */
export function keySequenceStartsWith(seq1: IKeySequence, seq2: IKeySequence): boolean {
  let keyCodes1 = seq1.keys.join();
  let keyCodes2 = seq2.keys.join();
  if (keyCodes1.length === 0 || keyCodes2.length === 0) {
    return false;
  }
  return keyCodes1.indexOf(keyCodes2) === 0 || keyCodes2.indexOf(keyCodes1) === 0;
}

/**
 * Converts a whole set of KeySequences into one keytip ID, which will be the ID for the last keytip sequence specified
 * keySequences should not include the initial keytip 'start' sequence
 * @param keySequences - Full path of IKeySequences for one keytip
 */
export function convertSequencesToKeytipID(keySequences: IKeySequence[]): string {
  let conversion = ktpPrefix;
  for (let keySequence of keySequences) {
    conversion += ktpSeparator + keySequence.keys.join(ktpSeparator);
  }
  return conversion;
}