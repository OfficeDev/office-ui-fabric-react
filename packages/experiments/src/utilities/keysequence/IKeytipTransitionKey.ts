import { KeytipTransitionModifier } from '../../Keytip';

export interface IKeytipTransitionKey {
  key: string;
  modifierKeys?: KeytipTransitionModifier[];
}

/**
 * Tests for equality between two IKeytipTransitionKeys
 *
 * @param key1 - First IKeytipTransitionKey
 * @param key2 - Second IKeytipTransitionKey
 * @returns {boolean} T/F if the transition keys are equal
 */
export function transitionKeysAreEqual(key1: IKeytipTransitionKey, key2: IKeytipTransitionKey): boolean {
  if (key1.key !== key2.key) {
    return false;
  }

  let mod1 = key1.modifierKeys;
  let mod2 = key2.modifierKeys;

  if ((!mod1 && mod2) || (mod1 && !mod2)) {
    // Not equal if one modifier is defined and the other isn't
    return false;
  }

  if (mod1 && mod2) {
    if (mod1.length !== mod2.length) {
      return false;
    }

    // Check that all items in mod1 are in mod2
    // Order doesn't matter
    let mod2Copy = [...mod2];
    for (let i = 0; i < mod1.length; i++) {
      let index = mod2Copy.indexOf(mod1[i]);
      if (index < 0) {
        return false;
      }
      mod2Copy.splice(index, 1);
    }
    if (mod2Copy.length !== 0) {
      return false;
    }
  }

  return true;
}

/**
 * Tests if 'key' is present in 'keys'
 *
 * @param keys - Array of IKeytipTransitionKey
 * @param key - IKeytipTransitionKey to find in 'keys'
 * @returns {boolean} T/F if 'keys' contains 'key'
 */
export function transitionKeysContain(keys: IKeytipTransitionKey[], key: IKeytipTransitionKey): boolean {
  for (let i = 0; i < keys.length; i++) {
    if (transitionKeysAreEqual(keys[i], key)) {
      return true;
    }
  }
  return false;
}
