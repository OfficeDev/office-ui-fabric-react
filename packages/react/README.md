# @uifabric/react

**React components for [Office UI Fabric React](https://dev.microsoft.com/fabric)**

These are not production-ready components and **should never be used in product**. This space is useful for testing new components whose APIs might change before final release.

To import React components:

```js
import { ComponentName } from '@uifabric/react';
```

Once the React component graduates to a production release, the component will be available at:

```js
import { ComponentName } from 'office-ui-fabric-react';
```
