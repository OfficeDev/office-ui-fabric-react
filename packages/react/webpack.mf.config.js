const path = require('path');
const fs = require('fs');
const webpack = require('webpack');
const getResolveAlias = require('@fluentui/scripts/webpack/getResolveAlias');
const { createServeConfig } = require('@fluentui/scripts/webpack/webpack-resources');
const { webpackMerge } = require('just-scripts');

const shared = {
  react: { singleton: true },
  'react-dom': { singleton: true },
};

const rootComponents = fs
  .readdirSync(path.join(__dirname, 'src'))
  .filter(filename => filename[0].toUpperCase() === filename[0])
  .map(filename => filename.replace(/\.tsx?/, ''));
const rootComponentsExposes = {};
for (const component of rootComponents) {
  rootComponentsExposes[`./lib/${component}`] = `@fluentui/react/src/${component}`;
}

// TODO: fork-ts-checker-webpack-plugin needs to be taught about the resolve alias -> pathmapping in order for the type checking to work
process.env.SKIP_TYPECHECK = true;

const myConfig = {
  output: {},
  entry: './empty-entrypoint.js',
  mode: 'development',
  devtool: 'cheap-module-source-map',
  optimization: {
    concatenateModules: false,
  },
  module: {
    rules: [],
  },
  plugins: [
    new webpack.container.ModuleFederationPlugin({
      name: 'fluentuiReact',
      filename: 'remoteEntry.js',
      exposes: {
        ...rootComponentsExposes,
        './lib/compat/Button': '@fluentui/react/src/compat/Button',
        '.': '@fluentui/react/src/index',
        './lib/index': '@fluentui/react/src/index',
      },
      shared,
    }),
  ],
  resolve: {
    alias: getResolveAlias(),
  },
  devServer: {
    port: 2345,
    static: './dist',
  },
};

const metaConfig = webpackMerge.merge(createServeConfig(), myConfig);

module.exports = metaConfig;
