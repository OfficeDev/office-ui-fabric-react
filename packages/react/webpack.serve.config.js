const path = require('path');
const resources = require('@uifabric/build/webpack/webpack-resources');

module.exports = resources.createServeConfig({
  entry: './src/demo/index.tsx',

  output: {
    filename: 'demo-app.js',
  },

  externals: {
    react: 'React',
    'react-dom': 'ReactDOM',
  },

  resolve: {
    alias: {
      '@fluentui/react/src': path.join(__dirname, 'src'),
      '@fluentui/react/lib': path.join(__dirname, 'lib'),
      '@fluentui/react': path.join(__dirname, 'lib'),
      'Props.ts.js': 'Props',
      'Example.tsx.js': 'Example',
    },
  },
});
