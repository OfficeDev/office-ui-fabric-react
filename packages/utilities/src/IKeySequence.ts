import { ktpPrefix, ktpSeparator } from './KeytipConstants';
import { addElementAtIndex } from './array';

export type IKeySequence = string;

/**
 * Tests for equality between two arrays of IKeySequence
 *
 * @param seq1 - First IKeySequence[]
 * @param seq2 - Second IKeySequence[]
 * @returns {boolean} T/F if the sequences are equal
 */
export function fullKeySequencesAreEqual(seq1: IKeySequence[], seq2: IKeySequence[]): boolean {
  if (seq1.length !== seq2.length) {
    return false;
  }

  for (let i = 0; i < seq1.length; i++) {
    if (seq1[i] !== seq2[i]) {
      return false;
    }
  }

  return true;
}

/**
 * Method returns true if the key squence with minimum length is in the other key sequence.
 * If the minium length is zero, then it will default to false.
 *
 * @param seq1 - First IKeySequence
 * @param seq2 - Second IKeySequence
 * @returns {boolean} T/F if seq1 starts with seq2, or vice versa
 */
export function keySequenceStartsWith(seq1: IKeySequence, seq2: IKeySequence): boolean {
  if (seq1.length === 0 || seq2.length === 0) {
    return false;
  }
  return seq1.indexOf(seq2) === 0 || seq2.indexOf(seq1) === 0;
}

/**
 * Converts a whole set of KeySequences into one keytip ID, which will be the ID for the last keytip sequence specified
 * keySequences should not include the initial keytip 'start' sequence
 *
 * @param keySequences - Full path of IKeySequences for one keytip
 * @returns {string} String to use for the keytip ID
 */
export function convertSequencesToKeytipID(keySequences: IKeySequence[]): string {
  return keySequences.reduce((prevValue: string, keySequence: IKeySequence): string => {
    return prevValue + ktpSeparator + keySequence.split('').join(ktpSeparator);
  }, ktpPrefix);
}

/**
 * Merges an overflow sequence with a key sequence
 *
 * @param keySequences - Full sequence for one keytip
 * @param overflowKeySequences - Full overflow keytip sequence
 * @returns {IKeySequence[]} Sequence that will be used by the keytip when in the overflow
 */
export function mergeOverflowKeySequences(keySequences: IKeySequence[], overflowKeySequences: IKeySequence[]): IKeySequence[] {
  const overflowSequenceLen = overflowKeySequences.length;
  const overflowSequence = [...overflowKeySequences].pop();
  const newKeySequences = [...keySequences];
  return addElementAtIndex(newKeySequences, overflowSequenceLen - 1, overflowSequence!);
}