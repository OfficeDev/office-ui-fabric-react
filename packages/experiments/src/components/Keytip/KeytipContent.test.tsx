import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { IKeySequence } from '../../utilities/keysequence/IKeySequence';
import { KeytipContent } from './KeytipContent';
const sequence: IKeySequence[] = [{ keys: ['a'] }];
const keyCont = 'A';

describe('Keytip', () => {
  it('renders Keytip correctly', () => {
    const componentContent = renderer.create(<KeytipContent content={ keyCont } keySequences={ sequence } />);
    let tree = componentContent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});