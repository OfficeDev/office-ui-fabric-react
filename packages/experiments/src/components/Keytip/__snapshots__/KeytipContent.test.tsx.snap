// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keytip renders Keytip correctly 1`] = `
<div
  className=

      {
        background-color: #212121;
        visibility: hidden;
      }
>
  <span
    className=

        {
          background-color: #212121;
          color: #eaeaea;
          display: inline-block;
          height: 17px;
          line-height: 17px;
          min-width: 11px;
          padding-left: 3px;
          padding-right: 3px;
          text-align: center;
        }
    id="ktp-a"
  >
    A
  </span>
</div>
`;
